Update records-wip.md (#3353)
dotnet/csharplang: proposals/records-wip.md

* A public get-only auto-property is created. Its value is initialized during construction with the value of the corresponding primary constructor parameter. Each "matching" inherited abstract property's get accessor is overridden.

>   * This property is also `initonly`, meaning the backing field can be modified in the [with](#With) expression below, if the corresponding `get` accessor is accessible

Reviewer: Perhaps 
```suggestion
  * This property also has an `init` accessor, meaning the backing field can be modified in the [with](#With) expression below, if the `init` accessor is accessible
```
given that we're going with `init` being the keyword?

Author: Out of scope for now, I'm just trying to make records work, not user-declared `initonly` accessors.

produce a copy of the receiver expression with modifications to properties
listed in the `anonymous_object_initializer`.

A valid `with` expression has a receiver with a non-void type. The receiver type must contain an accessible parameterless instance method called `Clone` whose return type must be the receiver type, or a base type thereof.

On the right hand side of the `with` expression is an `anonymous_object_initializer` with a
sequence of assignments with a compiler-generated record property (see Properties, above) of the receiver 
on the left-hand side of the assignment, and an arbitrary expression on the right-hand side which is 

> implicitly convertible to the type of the left-hand side.

Reviewer 2: >left-hand side [](start = 42, length = 14)

You mean the type of the property named on the left-hand-side?  (Only expressions have types)

Author: Yup! Fixed.

The evaluation of a `with` expression is equivalent to calling the `Clone` method exactly once,
and then setting the backing field of each record property in the argument list to its corresponding

> expression, in lexical order, using the result of the `Clone` method as the receiver.

Reviewer 2: >using the result of the `Clone` method as the receiver [](start = 30, length = 54)

The `Clone` method might not have that property as a member (in the case that it returns a base type).

Author: Fixed. The properties should in the return type of `Clone`, not the receiver type

produce a copy of the receiver expression with modifications to properties
listed in the `anonymous_object_initializer`.

> A valid `with` expression has a receiver with a non-void type. The receiver type must contain an accessible parameterless instance method called `Clone` whose return type must be the type of the receiver express type, or a base type thereof.

Reviewer 2: >express [](start = 204, length = 7)

What does "express" mean here?

parameterless instance method called `Clone` whose return type must be the receiver type, or a base type thereof.

On the right hand side of the `with` expression is an `anonymous_object_initializer` with a

> sequence of assignments, each with an `initonly` property (see [Properties](#Properties)) of the `Clone`

Reviewer: ```suggestion
sequence of assignments, each with an `init` property (see [Properties](#Properties)) of the `Clone`
```

return type on the left-hand side of the assignment, and an arbitrary expression on the right-hand side which is 
implicitly convertible to the type of the property.

The evaluation of a `with` expression calls the `Clone` method exactly once,

> and then sets the backing field of each `initonly` property in the argument list to 

Reviewer: ```suggestion
and then sets the backing field of each `init` property in the argument list to 
```
And perhaps it should be calling the `init` accessor instead of the backing field?

The evaluation of a `with` expression is equivalent to calling the `Clone` method exactly once,

> and then setting the backing field of each record property in the argument list to its corresponding

Reviewer 2: >backing field [](start = 21, length = 13)

The backing field is not typically accessible to the caller.

Author: It does if it's generated!

Reviewer 2: Not according to the spec above.  Per spec, an auto-property is generated, and there is no accessible backing field in that case.

---
In reply to: [406952450](https://github.com/dotnet/csharplang/pull/3353#discussion_r406952450) [](ancestors = 406952450)

Author: There is a backing field, and the semantics are defined st. it were as though the backing field were set. There's no requirement that the operation being performed is a legal C# assignment, so there's no requirement that the backing field is accessible.

Reviewer 2: You probably want to say some of this in the spec.

---
In reply to: [406956451](https://github.com/dotnet/csharplang/pull/3353#discussion_r406956451) [](ancestors = 406956451)

Author: > You probably want to say some of this in the spec. 

Fair! I'll try to clarify that where some of the existing language rules are being bypassed.

> Moreover, it isn't clear how this could be mapped onto verifiable IL.

IL verifiability isn't well-defined in the C# language specification, so I'll leave that for future work.

Reviewer 2: When the spec defines something as "equivalent to" something else, that includes legality of doing that something else in C#.  If instead you're defining the runtime semantics, just say that the backing field is assigned (no "equivalence" is involved).

---
In reply to: [406956451](https://github.com/dotnet/csharplang/pull/3353#discussion_r406956451) [](ancestors = 406956451)

Reviewer 2: As I understand the initonly feature, its backing field cannot be set directly by the caller. But its setter can be called during the initialization phase. If you intend the `with` expression to be part of the initialization phase, you need to specify where it begins, what value's initialization field it is the initialization phase *for*, and where it ends.

---
In reply to: [406950140](https://github.com/dotnet/csharplang/pull/3353#discussion_r406950140) [](ancestors = 406950140)

> and then setting the backing field of each record property in the argument list to its corresponding

Reviewer 2: >record property [](start = 43, length = 15)

I don't know what a "record property" is. Also, the return type of the `Clone` method might not be a type that has the named property as a member.

Author: Defined a record property above. The record property that is compiler-generated if one of equivalent signature does not exist.

Reviewer 2: There is no requirement that it be a record type. Moreover, I don't know what it means for a type to be a record type from the point of view outside the type itself.

---
In reply to: [406952410](https://github.com/dotnet/csharplang/pull/3353#discussion_r406952410) [](ancestors = 406952410)

Author: If it's not a record type, there are no generated record members. It's just a consequence of the precondition.

Reviewer 2: Are you saying that the "record"ness of a type and the "generated record property"ness of a property, are visible to the language externally and across compilation boundaries?  If so, I would expect to see that in the implementation as well.

---
In reply to: [406953742](https://github.com/dotnet/csharplang/pull/3353#discussion_r406953742) [](ancestors = 406953742)

Author: That would seem to be true! However, the implementation is a prototype, and only implements a subset of the spec 😄

Reviewer 2: I thought you were only specifying what you're implementing.

---
In reply to: [406956349](https://github.com/dotnet/csharplang/pull/3353#discussion_r406956349) [](ancestors = 406956349)

Author: Alternatively, I could phrase this as "the property is `init`-only, but I don't have a full specification there either, so it seems like either way we'll end up with some specification gap.

This phrasing makes sense... it's just broader than will eventually be implemented.
